styx-scheduler-service/src/main/java/com/spotify/styx/state/QueuedEventConsumer.java

```diff
@@ -0,0 +1,128 @@
+/*-
+ * -\-\-
+ * Spotify Styx Common
+ * --
+ * Copyright (C) 2016 - 2017 Spotify AB
+ * --
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ * 
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ * 
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ * -/-/-
+ */
+
+package com.spotify.styx.state;
+
+import com.spotify.styx.model.SequenceEvent;
+import com.spotify.styx.publisher.EventInterceptor;
+import java.io.IOException;
+import java.util.NoSuchElementException;
+import java.util.Objects;
+import java.util.Queue;
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.LinkedBlockingQueue;
+import java.util.concurrent.TimeUnit;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Single threaded asynchronous event consumer queue. It requires a {@link EventInterceptor}
```

Comment:
Lots of ad-hoc code when there seem to be primitives in Java for this... I'm looking at the snippet with `FinalizableDelegatedExecutorService` in https://stackoverflow.com/questions/30208786/run-only-one-thread-with-threadpoolexecutor